### ISSUE TITLE
[spi_device][hmac] TL outstanding request is 2

### ISSUE THREAD

**Issue by weicaiyang** [lowRISC/opentitan#860]
The buffer depth in socket is 1 for spi_dev and hmac, which makes the maximum outstanding request of TL to be 2. 
As discussed, we'd better set the buffer in IP to 0 and let xbar to handle buffering.

@cindychip since I added to check if max outstanding is 1. You may see some of you tests fail. You can add below in hmac_env_cfg::initialize to avoid the error for now.
`m_tl_agent_cfg.max_outstanding_req = 2;`

**Comment by tjaychen** [comment]
are you sure removing the buffering element on the design side is the right thing to do? Eventually we will probably run into IPs that intentionally buffer so that it can handle more transactions in-flight.  Wouldn't the right way be for DV / Design to somehow make this information discoverable? Sorry maybe that's the intent and I just missed it.

**Comment by cindychip** [comment]
And actually @weicaiyang  I ran a sequence and found 3 pending transactions: if you run `make TEST_NAME=hmac_reset SEED=7980`
Is there a way to turn off this checking? Because I think from the sequence stand of point, we will just stream in as many msg as design TLUL interface allows. I prefer to let HMAC pass the checklist and be stable at this point, rather than changing RTL to support this.

**Comment by weicaiyang** [comment]
> are you sure removing the buffering element on the design side is the right thing to do? Eventually we will probably run into IPs that intentionally buffer so that it can handle more transactions in-flight. Wouldn't the right way be for DV / Design to somehow make this information discoverable? Sorry maybe that's the intent and I just missed it.

I feel it's good to keep all devices consistent. But I'm also OK if spi_dev/hmac has more buffer as long as we clearly document it.

**Comment by weicaiyang** [comment]
> And actually @weicaiyang I ran a sequence and found 3 pending transactions: if you run `make TEST_NAME=hmac_reset SEED=7980`
> Is there a way to turn off this checking? Because I think from the sequence stand of point, we will just stream in as many msg as design TLUL interface allows. I prefer to let HMAC pass the checklist and be stable at this point, rather than changing RTL to support this.

setting  m_tl_agent_cfg.max_outstanding_req=0 should bypass the checking.

**Comment by tjaychen** [comment]
ah sorry, i don't mean we shouldnt remove it for spi / hmac.  For our purposes, there's no real reason for those to buffer.  I just meant there will be cases in the future where an IP will do so natively, and we probably want to find the right process to communicate that to DV, or somehow pass that information automatically to DV.

**Comment by cindychip** [comment]
HI Weicai, I do not think max_outstanding_req = 0 works, it gives
error "Response
queue overflow, response was dropped".
I think we still need to either find the max outstanding hmac can hold
(which I believe now can be 6), or change the tl_monitor code to disable
the checking. Can you provide the link to the doc that says max 1
outstanding?

**Comment by eunchan** [comment]
I just jumped in this issue. Thanks @cindychip @sriyerg for the detailed explanation.

From the designer point of view, extracting the number of outstanding value from the data path should be done but by manually. the tool (reggen) cannot extract the detailed #acceptance. But only can assume the lower bound of the # acceptances in the device.

But, this issue is really good for me to think about the regtool and its `window` feature. When regtool just got the window feature, `tlul_fifo_sync` has no logic for Depth := 0. So, I just set it to 1, which results in this kind of unexpected behavior. 

Now, thanks to @tjaychen , `tlul_fifo_sync` can have depth 0, which totally removes internal FIFOs inside `tl_socket_1n`. I will change the register template to make the Depth to 0.

This, however, doesn't mean that the tlul_window port has 1 outstanding acceptance. It still has more than that due to `tlul_adapter_sram` or any other logic to be implemented inside the design. But at least this gives more flexibility of FIFO depth control to the design .

I will make a PR , which results in HMAC to be D2 stage again, and probably re-trigger the regression of HMAC to gather the coverage report. (sorry @cindychip !)

CC: @sjgitty @aytong
